Refactor: Split INSERT and UPDATE entry points in Indexer

jeeminso opened the pull request:
## Summary of the changes / Why this improves CrateDB

Fixes https://github.com/crate/crate/issues/18219.

---

default/generated column semantics:
- default: insert: Generate if absent, otherwise use provided value
- default: update: Re-use existing value, unless there is an explicit new assignment; never re-generate, not even for nondeterministic.
- generated: insert: Generate if absent; validate if present, unless nondeterministic - in that case use as-is
- generated: update: Re-generate if absent; validate if present, unless nondeterministic - in that case use as-is.
- Additionally, non-deterministic values must be generated on the primary shard and sent to the replica, while deterministic values could also be generated on both primary and replica. Either approach would be fine, although regenerating would reduce traffic



## Checklist

 - [ ] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [ ] Updated documentation & `sql_features` table for user facing changes
 - [ ] Touched code is covered by tests
 - [ ] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

BaurzhanSakhariev commented:
Hi @jeeminso, as agreed also took a look.

Second commit - Integrated Jordi's change as is, generated works, default is still null
Third commit - fix default  = null
Fourth commit - don't update undeterministic default on updates

```
Current failure: java.lang.AssertionError: [Update must have re-evaluated the generated expression] 
Expecting actual:
  1754658885232L
to be greater than:
  1754658885232L
```
  
Need to leave, will check later - feel free to fix it if you already have an idea

jeeminso commented:
Thanks for your input Baur, I will try to complete the refactoring first - separating insert and update execution paths(I believe this is Jordi's suggestion) then will try to merge your fix back in.

jeeminso commented:
retest this please

jeeminso commented on server/src/main/java/io/crate/execution/dml/Indexer.java:
```diff
@@ -1222,15 +1278,21 @@ public IndexItem convert(Doc doc, Symbol[] updateAssignments, Object[] excludedV
                 } else if (ref instanceof GeneratedReference genRef && !genRef.isDeterministic()) {
                     insertValues[i] = null;
-                } else if (ref.defaultExpression() != null && !ref.defaultExpression().isDeterministic()) {
-                    insertValues[i] = null;
```
update must not re-evaluate nondeterministic defaults

jeeminso commented:
Status update - the first four commits are small incremental refactorings to split execution paths for `insert` and `update`. Afterwards, port fixes from https://github.com/crate/crate/pull/18251, added integration tests for `update`, `insert`, `insert-on-conflict` and fixed the failures from the new tests. Currently failures only happen for `insert-on-conflict` scenarios. As shown in the last commit and as discussed, I am trying to use two Indexer instances for `insert-on-conflict` queries.

Once `insert-on-conflict` is addressed, I can expand on the tests for more edge cases. Also the tests can be reduced in the end, I must have introduced some redundant tests.

matriv commented on server/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java:
```diff
@@ -2065,6 +2043,34 @@ ON CONFLICT (id) DO UPDATE SET value = excluded.value
+            .isGreaterThan(writeTs);
+
+        assertThat(updatedZ)
+            .as("Undeterministic default value is not re-evaluated")
```
I'm confused with this, the updatedZ shouldn't be re-evaluated, right?
(that's why you assert it against previous value `z`)

jeeminso commented:
continuing on https://github.com/crate/crate/pull/18323.
